### PR TITLE
add gpu resource driver helm chart

### PR DESCRIPTION
**Pull request by oxxenix**
image: intel/intel-gpu-resource-driver:v0.2.0
crds files are added under crds folder

### CODE REVIEW

File: charts/intel-gpu-resource-driver/templates/gpu-resource-driver.yaml [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+          valueFrom:
+            fieldRef:
+              fieldPath: metadata.namespace
+        securityContext:
+          {{- toYaml .Values.kubeletPlugin.containers.init.securityContext | nindent 10 }}
+      containers:
+      - name: kubelet-plugin
+        image: {{ include "intel-gpu-resource-driver.fullimage" . }}
+        imagePullPolicy: {{ .Values.image.pullPolicy }}
+        command: ["/kubelet-plugin", "-v", "5"]
```
Comment (byako) [inline charts/intel-gpu-resource-driver/templates/gpu-resource-driver.yaml:124]:
```suggestion
        command: ["/kubelet-plugin"]
```
Let's not enable such verbosity by default.

File: charts/intel-gpu-resource-driver/templates/gpu-resource-driver.yaml [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+      labels:
+        {{- include "intel-gpu-resource-driver.templateLabels" . | nindent 8 }}
+    spec:
+      serviceAccount: intel-gpu-resource-driver-service-account
+      serviceAccountName: {{ include "intel-gpu-resource-driver.serviceAccountName" . }}
+      containers:
+      - name: controller
+        image:  {{ include "intel-gpu-resource-driver.fullimage" . }}
+        imagePullPolicy: {{ .Values.image.pullPolicy }}
+        command: ["/controller", "-v", "6"]
```
Comment (byako) [inline charts/intel-gpu-resource-driver/templates/gpu-resource-driver.yaml:216]:
```suggestion
        command: ["/controller"]
```
Let's not enable such verbosity by default.

File: charts/intel-gpu-resource-driver/values.yaml [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+selectorLabelsOverride: {}
+
+crds:
+  enabled: true
+
+imagePullSecrets: []
+image:
+  repository: intel
+  name: intel-gpu-resource-driver
+  pullPolicy: Always
```
Comment (byako) [inline charts/intel-gpu-resource-driver/values.yaml:14]:
```suggestion
  pullPolicy: IfNotPresent
```
For development purposes pullPolicy `Always` is good, but for release we don't need the same image pulled all the time, let it be "IfNotPresent"

**Review by byako — CHANGES_REQUESTED** [review]
According to your previous pull request in this repo (#43 ), a License file is needed in this chart ?

**Comment by tkatila** [comment]
> > According to your previous pull request in this repo (#43 ), a License file is needed in this chart ?
> 
> @tkatila, @poussa do we need to add licenses for all Intel helm charts?

Most of the charts do not have it. But I don't think it hurts to have it.

**Comment by eero-t** [comment]
Looks fine, but I wonder why YAML file contents are split differently from the project this is installing:
 https://github.com/intel/intel-resource-drivers-for-kubernetes/tree/main/deployments/

Here class + parameters, defaults, and daemonset+deployment are in same file, although in driver repo they are all 3 in separate YAML files.  Whereas RBAC stuff is here in separate files but in driver repo in same file with deployments.

Wouldn't it be easier to check for changes and sync them, if k8s objects would be split similarly to separate files (and be in same order in them)?

**Comment by oxxenix** [comment]
> Looks fine, but I wonder why YAML file contents are split differently from the project this is installing: https://github.com/intel/intel-resource-drivers-for-kubernetes/tree/main/deployments/
> 
> Here class + parameters, defaults, and daemonset+deployment are in same file, although in driver repo they are all 3 in separate YAML files. Whereas RBAC stuff is here in separate files but in driver repo in same file with deployments.
> 
> Wouldn't it be easier to check for changes and sync them, if k8s objects would be split similarly to separate files (and be in same order in them)?

I've been relying on the Helm chart available at [dra-example-driver](https://github.com/kubernetes-sigs/dra-example-driver/tree/main/deployments/helm/dra-example-driver/templates), which separates RBAC files and separately manages the deployments of the controller and kubelet-plugin.

I'm flexible and open to adapting to any upcoming changes. If the preference is to maintain all files individually, I'm ready to align with that approach.

File: charts/intel-gpu-resource-driver/values.yaml [hunk 4]
```diff
@@ -0,0 +1,85 @@
+# Default values for intel-gpu-resource-driver.
+nameOverride: ""
+namespaceOverride: ""
+fullnameOverride: ""
+selectorLabelsOverride: {}
+
+crds:
+  enabled: true
```
Comment (tkatila) [inline charts/intel-gpu-resource-driver/values.yaml:8]: Is this used? AFAIK, CRDs cannot be disabled when they are defined under 'crds' directory in Helm.

File: charts/intel-gpu-resource-driver/templates/gpu-resource-driver.yaml [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+    type=gpu
+    shareable=true
+  vf-memory.config: |
+    {
+      "max1550": 16385,
+      "max1450": 16384,
+      "max1100": 8192,
+      "flex140": 2048,
+      "flex170": 4096,
+    }
```
Comment (tkatila) [inline charts/intel-gpu-resource-driver/templates/gpu-resource-driver.yaml:80]: Are these something that should be configurable via values?

Comment (byako) [reply charts/intel-gpu-resource-driver/templates/gpu-resource-driver.yaml:80]: Not necessarily, this ConfigMap is read often, so once cluster admin changes it, there's even no need to redeploy anything, so can be done post-installation.

**Comment by eero-t** [comment]
> I've been relying on the Helm chart available at [dra-example-driver](https://github.com/kubernetes-sigs/dra-example-driver/tree/main/deployments/helm/dra-example-driver/templates), which separates RBAC files and separately manages the deployments of the controller and kubelet-plugin.
> 
> I'm flexible and open to adapting to any upcoming changes. If the preference is to maintain all files individually, I'm ready to align with that approach.

I would think diffing / updating them would be easier if k8s objects in them are in same order and split to files with same names as in the original [deployments](https://github.com/intel/intel-resource-drivers-for-kubernetes/tree/main/deployments/) for the driver it's installing.  That could be separate PR though.

@byako, any comments on what you prefer?

File: charts/intel-gpu-resource-driver/values.yaml (hunk 4 above)
Comment (oxxenix) [reply charts/intel-gpu-resource-driver/values.yaml:8]: If crds.enabled is set to` true,` it usually means the CRDs in the crds folder are installed or managed with the Helm chart. If it's `false`, the CRDs are not managed by the Helm chart, useful when they're already installed or managed independently.

Comment (tkatila) [reply charts/intel-gpu-resource-driver/values.yaml:8]: CRDs are not templated so I think having the value here doesn't apply. Based on the documentation, CRD installation can only be skipped from command line:
https://helm.sh/docs/chart_best_practices/custom_resource_definitions/#method-1-let-helm-do-it-for-you

Comment (eero-t) [reply charts/intel-gpu-resource-driver/values.yaml:8]: And according to Helm (v3) documentation, even more worrying is that Helm just will not touch already installed CRDs, regardless of whether they would need to be upgraded or deleted.   There's no option to upgrade obsolete CRDs with Helm.

Best summary I found from the various Helm tickets on the subject, was this: https://github.com/helm/helm/issues/10585#issuecomment-1109561457

Comment (oxxenix) [reply charts/intel-gpu-resource-driver/values.yaml]: The `crds.enabled: true` setting in the values.yaml file of a Helm chart is there to make sure all the necessary CRDs are automatically installed for users, especially for those who might not be very familiar with CRDs. If you set it to `false`, it tells the chart not to install or update CRDs, which is useful for people who prefer to handle CRDs themselves. Normally, CRDs are stored in a `crds/ `folder in the chart and are installed by default unless you choose to skip them. This setting gives you more control over that process.

You can see a similar setup in the [kube-prometheus-stack chart](https://github.com/prometheus-community/helm-charts/blob/ee637d09aa6b57e4f0ffcf4d5488c34ca11ff10a/charts/kube-prometheus-stack/values.yaml#L31C1-L34C16).

For our chart, I removed this line and added a note in the README.md that CRDs should be installed separately

**Review by eero-t — APPROVED** [review]
Looks fine for me, but this is more Alexey's and Tuomas' field I think.

File: charts/intel-gpu-resource-driver/values.yaml (hunk 4 above)
Comment (tkatila) [reply charts/intel-gpu-resource-driver/values.yaml]: kube-prometheus-stack chart has a more complex setup. They have a dependency to a subchart that only includes the CRDs: https://github.com/prometheus-community/helm-charts/tree/ee637d09aa6b57e4f0ffcf4d5488c34ca11ff10a/charts/kube-prometheus-stack/charts/crds
And that is enabled or disabled based on the Chart.yaml's conditions based on the values (crds.enabled):
https://github.com/prometheus-community/helm-charts/blob/ee637d09aa6b57e4f0ffcf4d5488c34ca11ff10a/charts/kube-prometheus-stack/Chart.yaml#L44

I think that's a nice way to overcome the restrictions on Helm. In any case, I think it's fine now that the `crds.enabled` is removed.

**Review by tkatila — APPROVED** [review]

**Review by byako — APPROVED** [review]
LGTM

**Pull request merged**